File: stack.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

void waitForEnter() {
    printf("\nPress Enter to continue...");
    int c;
    while ((c = getchar()) != '\n' && c != EOF); 
    getchar(); 
}

void push(struct Node** top, int data, int showMessage); 

void saveToFile(struct Node* top) {
    FILE *file = fopen("stack_data.txt", "w");
    if (file == NULL) return;
    
    struct Node* temp = top;
    while (temp != NULL) {
        fprintf(file, "%d\n", temp->data);
        temp = temp->next;
    }
    fclose(file);
}

void loadFromFile(struct Node** top) {
    FILE *file = fopen("stack_data.txt", "r");
    if (file == NULL) return; 
    
    int data;
    int tempArr[1000]; 
    int count = 0;
    int i; 
    
    while (fscanf(file, "%d", &data) == 1) {
        tempArr[count] = data;
        count++;
    }
    
    for (i = count - 1; i >= 0; i--) {
        push(top, tempArr[i], 0); 
    }
    fclose(file);
}

void createStack(struct Node** top) {
    if (*top != NULL) {
        printf("\n[Notice] A stack already exists! Please use Option 3 to Push more items.\n");
        return;
    }

    char resp;
    int data;

    do {
        printf("Enter data to push: ");
        
        if (scanf("%d", &data) != 1) {
            printf("  -> [Error] Invalid input! Letters/characters are not allowed.\n");
            printf("\nReturning to main menu...\n");
            return; 
        }
        
        push(top, data, 0); 

        while (1) {
            printf("Add another node [Y/N]? ");
            scanf(" %c", &resp); 
            
            if (resp == 'Y' || resp == 'y' || resp == 'N' || resp == 'n') {
                break; 
            } else {
                printf("  -> [Error] Invalid input. Please enter 'Y' for Yes or 'N' for No.\n");
            }
        }
        
    } while (resp == 'Y' || resp == 'y');
    
    printf("\n[Success] Initial Stack created!\n");
}

void display(struct Node* top) {
    if (top == NULL) {
        printf("\nError: The stack is empty!\n");
        return;
    }

    struct Node* temp = top;
    
    printf("\n--- Current Stack ---\n");
    
    while (temp != NULL) {
        printf("[%d]\n", temp->data);
        temp = temp->next;
    }
    printf("-------------------\n");
}

void push(struct Node** top, int data, int showMessage) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    if (newNode == NULL) {
        printf("\nError: Memory allocation failed!\n");
        return;
    }
    
    newNode->data = data;
    newNode->next = *top;
    *top = newNode;
    
    if (showMessage) {
        printf("\n[Success] %d was pushed to the top of the stack!\n", data);
    }
}

void pop(struct Node** top) {
    if (*top == NULL) {
        printf("\nError: Stack Underflow! The stack is already empty.\n");
        return;
    }
    
    struct Node* temp = *top;
    int poppedData = temp->data;
    
    *top = (*top)->next;
    free(temp);
    
    printf("\n[Success] %d was popped from the stack!\n", poppedData);
}

void peek(struct Node* top) {
    if (top == NULL) {
        printf("\nError: The stack is empty!\n");
    } else {
        printf("\n[Peek] The item at the TOP of the stack is: %d\n", top->data);
    }
}

int main() {
    struct Node* top = NULL; 
    int choice, data;

    loadFromFile(&top);

    while (1) {
        system("cls"); 
        
        printf("===== Stack Implementation by Group 1 =====\n");
        printf("1. Create stack\n");
        printf("2. Display stack\n");
        printf("3. Push (insert item)\n");
        printf("4. Pop (remove item)\n");
        printf("5. Peek (view top item)\n");
        printf("6. Exit\n");
        printf("Enter your choice: ");
        
        if (scanf("%d", &choice) != 1) {
            printf("\n[Error] Invalid input! Letters/symbols are not allowed.\n");
            waitForEnter(); 
            continue; 
        }

        switch (choice) {
            case 1:
                createStack(&top);
                waitForEnter();
                break;
            case 2:
                display(top);
                waitForEnter();
                break;
            case 3:
                printf("Enter data to push onto the stack: ");
                if (scanf("%d", &data) != 1) {
                    printf("\n[Error] Invalid input! Please enter a number.\n");
                } else {
                    push(&top, data, 1); 
                }
                waitForEnter();
                break;
            case 4:
                pop(&top);
                waitForEnter();
                break;
            case 5:
                peek(top);
                waitForEnter();
                break;
            case 6:
                printf("\nSaving data and exiting program...\n");
                saveToFile(top);
                
                while (top != NULL) {
                    struct Node* temp = top;
                    top = top->next;
                    free(temp);
                }
                exit(0);
            default:
                printf("\n[Error] Invalid selection.\n");
                waitForEnter();
        }
    }
    return 0;
}

File: stack.py
from os import system
from pathlib import Path

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def waitForEnter():
    print("\nPress Enter to continue...")
    input()

def saveToFile(top):
    with open("stack_data.txt", "w") as file:
        temp = top
        while temp is not None:
            file.write(f"{temp.data}\n")
            temp = temp.next

def loadFromFile(top):
    if not Path("stack_data.txt").exists():
        return

    tempArr = []
    count = 0

    with open("stack_data.txt", "r") as file:
        for line in file:
            data = int(line.strip())
            tempArr.append(data)
            count += 1

    i = count - 1
    while i >= 0:
        push(top, tempArr[i], 0)
        i -= 1

def createStack(top):
    if top[0] is not None:
        print("\n[Notice] A stack already exists! Please use Option 3 to Push more items.")
        return

    while True:
        data = input("Enter data to push: ")

        if not data.lstrip("-").isdigit():
            print("  -> [Error] Invalid input! Letters/characters are not allowed.")
            print("\nReturning to main menu...\n")
            return

        push(top, int(data), 0)

        while True:
            resp = input("Add another node [Y/N]? ")

            if resp in ["Y", "y", "N", "n"]:
                break
            else:
                print("  -> [Error] Invalid input. Please enter 'Y' for Yes or 'N' for No.")

        if resp in ["N", "n"]:
            break

    print("\n[Success] Initial Stack created!")

def display(top):
    if top is None:
        print("\nError: The stack is empty!")
        return

    temp = top

    print("\n--- Current Stack ---")
    while temp is not None:
        print(f"[{temp.data}]")
        temp = temp.next
    print("-------------------")

def push(top, data, showMessage):
    newNode = Node(data)
    newNode.next = top[0]
    top[0] = newNode

    if showMessage:
        print(f"\n[Success] {data} was pushed to the top of the stack!")

def pop(top):
    if top[0] is None:
        print("\nError: Stack Underflow! The stack is already empty.")
        return

    temp = top[0]
    poppedData = temp.data

    top[0] = top[0].next

    print(f"\n[Success] {poppedData} was popped from the stack!")

def peek(top):
    if top[0] is None:
        print("\nError: The stack is empty!")
    else:
        print(f"\n[Peek] The item at the TOP of the stack is: {top[0].data}")

def main():
    top = [None]
    loadFromFile(top)

    while True:
        system("cls")

        print("===== Stack Implementation by Group 1 =====")
        print("1. Create stack")
        print("2. Display stack")
        print("3. Push (insert item)")
        print("4. Pop (remove item)")
        print("5. Peek (view top item)")
        print("6. Exit")
        choice = input("Enter your choice: ")

        if not choice.isdigit():
            print("\n[Error] Invalid input! Letters/symbols are not allowed.")
            waitForEnter()
            continue

        choice = int(choice)

        if choice == 1:
            createStack(top)
            waitForEnter()

        elif choice == 2:
            display(top[0])
            waitForEnter()

        elif choice == 3:
            data = input("Enter data to push onto the stack: ")
            if not data.lstrip("-").isdigit():
                print("\n[Error] Invalid input! Please enter a number.")
            else:
                push(top, int(data), 1)
            waitForEnter()

        elif choice == 4:
            pop(top)
            waitForEnter()

        elif choice == 5:
            peek(top)
            waitForEnter()

        elif choice == 6:
            print("\nSaving data and exiting program...")

            saveToFile(top[0])

            while top[0] is not None:
                temp = top[0]
                top[0] = top[0].next
                del temp

            exit()

        else:
            print("\n[Error] Invalid selection.")
            waitForEnter()

main()
